rllib/env/multi_agent_env_runner.py

```diff
@@ -0,0 +1,887 @@
+        self._states = {agent_id: None for agent_id in self.marl_module.keys()}
+
+    @override(EnvRunner)
+    def sample(
+        self,
+        *,
+        num_timesteps: int = None,
+        num_episodes: int = None,
+        explore: bool = True,
+        random_actions: bool = False,
+        with_render_data: bool = False,
+    ) -> List["MultiAgentEpisode"]:
+        """Runs and returns a sample (n timesteps or m episodes) on the env(s).
+
+        Args:
+            num_timesteps: int. Number of timesteps to sample during rollout.
+                Note, only one parameter, `num_timetseps` or `num_episodes`
+                can be provided.
+            num_episodes: int. Number of episodes to sample during rollout.
+                Note, only one parameter, `num_timetseps` or `num_episodes`
+                    can be provided.
+            explore: boolean. If in exploration or inference mode. Exploration
+                mode might for some algorithms provide extza model outputs that
+                are redundant in inference mode.
+            random_actions: boolean. If actions should be sampled from the action
+                space. In default mode (i.e. `False`) we sample actions frokm the
+                policy.
+            with_render_data: If render data from the environment should be collected.
+                This is only available when sampling episodes, i.e. `num_episodes` is
+                not `None`.
+        Returns:
+            `Lists of `MultiAgentEpisode` instances, carrying the collected sample data.
+        """
+        assert not (num_timesteps is not None and num_episodes is not None)
+
+        # If no execution details are provided, use the config.
+        if num_timesteps is None and num_episodes is None:
+            if self.config.batch_mode == "truncate_episodes":
+                num_timesteps = self.config.get_rollout_fragment_length(
+                    worker_index=self.worker_index,
+                )
+            else:
+                num_episodes = 1
+
+        # Sample n timesteps
+        if num_timesteps is not None:
+            return self._sample_timesteps(
+                num_timesteps=num_timesteps,
+                explore=explore,
+                random_actions=random_actions,
+                force_reset=False,
+            )
+        # Sample m episodes.
+        else:
+            return self._sample_episodes(
+                num_episodes=num_episodes,
+                explore=explore,
+                random_actions=random_actions,
+                with_render_data=with_render_data,
+            )
+
+    def _sample_timesteps(
+        self,
+        num_timesteps: int,
+        explore: bool = True,
+        random_actions: bool = False,
+        force_reset: bool = False,
+    ) -> List["MultiAgentEpisode"]:
+        """Helper method to sample n timesteps.
+
+        Args:
+            num_timesteps: int. Number of timesteps to sample during rollout.
```

Comment:
See above: Let's enhance this docstring here explaining the different between the different `count_steps_by` settings.